allow load balancing mode to be set via env var

lloesche opened the pull request:
new:
- allows haproxy load balancing mode to be set via ENV var
e.g. HAPROXY_0_MODE=http
Instead of guessing the mode based on whether a hostname was supplied or not.

fixes:
- bind to http port if redirect_http_to_https is turned on

misc cleanup:
- empty lines
- prefix all setter functions with set_

lloesche commented:
I asked @cmaloney to review. He wrote a lot of the original code.

cmaloney commented:
While the proxying switches between HTTP and TCP mode, I don't think forcing it with an environment variable is actually very useful. Mainly it makes more opportunities for people to mess up / break config without warning or notice (Unless they happen to scan the servicerouter logs, which an app developer wouldn't see / do, and likely wouldn't have access to ssh to those hosts). As much as possible I want to avoid needing to do "checking that a set of options are compatible". Also I want things to "just work". Forwarding HTTP traffic as TCP will just work. The switch could be a little disconcerting for people, but documenting that in Virtual Hosts / getting developers to think through "virtual hosts uses http request fields so things must be treated as http".

lloesche commented:
> Forwarding HTTP traffic as TCP will just work

Nope, that was the problem. If we forward it as TCP then all X-Forwarded-For, X-Forwarded-Proto as well as http->https redirect doesn't work. Without any error message from haproxy whatsoever. 

I ran into that on the mesosphere accounts system where I set up https offloading and http2https redirect via servicerouter and haproxy would simply accept connections both unencrypted and encrypted. Problem turned out to be that mode was set to tcp so haproxy didn't add/modify the headers accordingly. That's why I added this option.

If you check line 297, it's choosing tcp as default if mode is not explicitly set by the user. So I just moved the else case to be the object's default for that property.

cmaloney commented:
Seems reasonable enough for now. Not what I feel is an ideal fix, but I'm also not strongly attached to it.

lloesche commented:
@cmaloney thanks!
